rs/backend/src/perf.rs

```
@@ -0,0 +1,82 @@
+pub struct PerformanceCount {
+    timestamp_ns_since_epoch: u64,
+    name: String,
+    instruction_count: u64,
+}
+impl PerformanceCount {
+    pub fn new(name: &str) -> Self {
+        let name = name.to_string();
+        let timestamp_ns_since_epoch = crate::time::time();
+        let instruction_count = crate::perf::raw_instruction_count();
+        PerformanceCount {
+            timestamp_ns_since_epoch,
+            name,
+            instruction_count,
+        }
+    }
+}
+
+/// Storage for recent performance counter snapshots.
+#[derive(CandidType, Deserialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct PerformanceCounts {
+    pub instruction_counts: VecDeque<PerformanceCount>,
+}
+
+impl PerformanceCounts {
+    /// Saves a performance counter snapshot.
+    pub fn save_instruction_count(&mut self, count: PerformanceCount) {
+        if self.instruction_counts.len() >= 100 {
+            self.instruction_counts.pop_front();
+        }
+        self.instruction_counts.push_back(count);
+    }
+
+    /// Generates sample data for use in tests
+    #[cfg(test)]
+    pub fn test_data() -> Self {
+        let mut ans = PerformanceCounts::default();
+        ans.save_instruction_count(PerformanceCount::default());
+        ans.save_instruction_count(PerformanceCount {
+            timestamp_ns_since_epoch: 999,
+            name: "Nein".to_string(),
+            instruction_count: 1,
+        });
+        ans.save_instruction_count(PerformanceCount::default());
+        ans
+    }
+}
+
+impl StableState for PerformanceCounts {
+    /// Try to serialize state. On error, return an empty vector.
+    fn encode(&self) -> Vec<u8> {
+        Candid((self,)).into_bytes().unwrap_or_default()
+    }
+
+    /// Parse performance counts.  On error, return a blank new structure.
+    fn decode(bytes: Vec<u8>) -> Result<Self, String> {
+        let (ans,) = Candid::from_bytes(bytes).map(|c| c.0).unwrap_or_default();
```

Since we're swallowing the error, should we log something if decoding fails?

---

If logging is wanted I'd rather return the string as an error and handle the error, or log it, further downstream.  I don't see the value in logging this.  If we care, it's because we are monitoring cycles over upgrades, in which case it is obvious that the pre-upgrade data has been dropped anyway.